File: BOJ/11053/A.py
import sys

sys.stdin = open(
    "/Users/sungwonlee/Documents/Projects/algorithms-practice/BOJ/11053/input.txt", 'r')

input = sys.stdin.readline

N = int(input().rstrip())
A = input().rstrip().split()
arr = [int(A[i]) for i in range(len(A))]

cnts = [-1 for i in range(len(arr))]
cnts[0] = 1
max_cnt = 1
for i in range(1, len(arr)):
    current_max_cnt = 0
    for j in range(i):
        if arr[j] < arr[i] and cnts[j] > current_max_cnt:
            current_max_cnt = cnts[j]
    cnts[i] = current_max_cnt + 1
    if current_max_cnt+1 > max_cnt:
        max_cnt = current_max_cnt+1
print(max_cnt)

File: BOJ/11053/A.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> arr;
vector<int> cnts(1001, 1);
int main()
{
    cin.tie(0);
    ios::sync_with_stdio(0);
    freopen("input.txt", "rt", stdin);
    int N;
    cin >> N;
    int ans = 1;
    for (int i = 0; i < N; i++)
    {
        int temp;
        cin >> temp;
        arr.push_back(temp);
    }
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < i; j++)
        {
            if (arr[j] < arr[i] && cnts[j] >= cnts[i])
            {

                cnts[i] = cnts[j] + 1;
            }
        }
    }
    for (int i = 0; i < N; i++)
    {
        int current = cnts[i];
        if (current > ans)
        {
            ans = current;
        }
    }
    cout << ans;
}
